designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+* A plug-able provider interface that can be used to develop off-tree providers. Off-tree provider development requires much broader thought such as how our API should be structured to enable API extension, how to version the APIs and ensure they function together, and how to build and distribute artifacts specific to a given provider.
+* Refactoring the mechanisms for workflow checkpoint-recovery. The interfaces for checkpoint-recovery interaction with workflows may require adjusting but the fundamentals of the design should not change.
+
+## Proposal
+
+The following sections describe smaller proposals that aim to better define concerns for workflows and providers, provide mechanisms for hooking into or extending workflows, and reduce cognitive load when understanding workflow execution. Some sections work together with others and some are more independent.
+
+These proposals also invokes a fundamental change in how we think about providers. Specifically, providers have historically been modeled as data providers for centralized orchestration logic. The data provider approach is built on assumptions around how providers are composed with a desire to treat everything the same; this has since been proven false. This proposal seeks to change providers to take on greater functional responsibility such as installation of provider specific components to an environment. This sets into motion an expectation that providers will handle their specifics with an orchestration layer telling them when to run.
+
+### Separation of Management and Workload workflows
+
+We propose separating management and workload cluster workflows and constraining tasks to only those required for cluster management operations. Additional tasks supported by EKS-A will be moved to an extensible post-workflow hook that we can continually build on to add more features separate from the minimal cluster provisioning tasks.
+
+Having implemented this change, there is an expectation that workflow maintainers follow a documented set of guidelines on the granularity of tasks and what tasks are appropriate for featuring explicitly in a workflow.
+
+#### Management cluster workflows
+
+Management cluster workflow will comprise of the following tasks.
+
+##### Create
+
+1. **Pre Create** executes pre workflow hooks.
+2. **Create Bootstrap Cluster** produces a bootstrap cluster that is used to provision the permanent management cluster. It should populate the workflow context with information to contact the bootstrap cluster.
+3. **Install Bootstrap Components** install CAPI components, EKS-A components, and provider specific components on a bootstrap cluster. Providers are responsible for installing the components they require.
```

Comment:
In the past, we have had to introduce functions in the provider interface to install things on the bootstrap cluster before installing CAPI components (example: PreCAPIInstall). How do we plan on handling that?

Reply:
See the section below on how hooks would be implemented. In summary, anything needing to run something pre or post a task would be bale to bind a pre or post hook to do that.